1. Активность пользователей зависит от дня недели. Причём в Москве и Петербурге это проявляется по-разному.

2. Утром в понедельник в Москве преобладают одни жанры музыки, а в Петербурге — другие. Это верно и для вечера пятницы.

3. Москва и Петербург предпочитают разные жанры музыки. В Москве чаще слушают поп-музыку, в Петербурге — русский рэп.

In [1]:
import pandas as pd
df = pd.read_csv('C:\\Users\\rivie\\Desktop\\yandex_music_project.csv')
df.columns

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')

In [2]:
df = df.rename(
 columns={
'  userID': 'user_id',
'Track': 'track',
'  City  ': 'city',
'Day':'day',
 }
)
df.columns #переименовываем названия столбцов к одному стилю 

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')

In [3]:
df.head(10)

user_id                        track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                    Преданная         IMPERVTOR  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE     И вновь продолжается бой               NaN  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

               city      time        day  
0  Saint-Petersburg  20:28:33  Wednesday  
1            Moscow  14:07:09     Friday  
2  Saint-Petersburg  20:58:07  Wednesday  
3  Saint-Petersburg  08:37:09     Monday  
4            Moscow  08:34:34     Monday  
5  Saint-Petersburg  13:09:41     Friday  
6            Moscow  13:00:07  Wednesday  
7            Moscow  20:47:49  Wednesday  
8            Moscow  09:17:40     Friday  
9  Saint-Petersburg  21:20:49  Wednesday

In [4]:
df.duplicated().sum() #проверяем файл на наличие полных дубликатов

3826

In [5]:
df = df.drop_duplicates().reset_index(drop=True) #удаляем дубликаты 

In [6]:
df.isna().sum() #ищем пустые значения

user_id       0
track      1155
artist     6772
genre      1127
city          0
time          0
day           0
dtype: int64

In [7]:
columns_to_replace = ['track','artist','genre']
for column in columns_to_replace:
    df[column] = df[column].fillna('unknown') #заполняем пустые значения

In [8]:
df.isna().sum()

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

In [9]:
sorted_df = df['genre'].sort_values()
sorted_df = sorted_df.unique()
sorted_df

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'alternativepunk', 'ambient', 'americana',
       'animated', 'anime', 'arabesk', 'arabic', 'arena',
       'argentinetango', 'art', 'audiobook', 'author', 'avantgarde',
       'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass',
       'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks',
       'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean',
       'caucasian', 'celtic', 'chamber', 'chanson', 'children', 'chill',
       'chinese', 'choral', 'christian', 'christmas', 'classical',
       'classicmetal', 'club', 'colombian', 'comedy', 'conjazz',
       'contemporary', 'country', 'cuban', 'dance', 'dancehall',
       'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr',
       'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo',
       'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic',
       'electropop', 'emo', 'entehno', '

In [10]:
wrong_genres = ['hip', 'hop', 'hip-hop']
for genre in df['genre']:
    for wrong_genre in wrong_genres:
        if genre == wrong_genre:
            df['genre'] = df['genre'].replace(genre, 'hiphop')

In [11]:
sorted_df = df['genre'].sort_values()
sorted_df = sorted_df.unique()
sorted_df

array(['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans',
       'alternative', 'alternativepunk', 'ambient', 'americana',
       'animated', 'anime', 'arabesk', 'arabic', 'arena',
       'argentinetango', 'art', 'audiobook', 'author', 'avantgarde',
       'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass',
       'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks',
       'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean',
       'caucasian', 'celtic', 'chamber', 'chanson', 'children', 'chill',
       'chinese', 'choral', 'christian', 'christmas', 'classical',
       'classicmetal', 'club', 'colombian', 'comedy', 'conjazz',
       'contemporary', 'country', 'cuban', 'dance', 'dancehall',
       'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr',
       'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo',
       'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic',
       'electropop', 'emo', 'entehno', '

## 1.Активность пользователей зависит от дня недели. Причём в Москве и Петербурге это проявляется по-разному.

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61253 entries, 0 to 61252
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  61253 non-null  object
 1   track    61253 non-null  object
 2   artist   61253 non-null  object
 3   genre    61253 non-null  object
 4   city     61253 non-null  object
 5   time     61253 non-null  object
 6   day      61253 non-null  object
dtypes: object(7)
memory usage: 3.3+ MB


In [13]:
df.head()

user_id                        track            artist  genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile   rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg   rock   
2    20EC38            Funiculì funiculà       Mario Lanza    pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice   folk   
4  E2DC1FAE                  Soul People        Space Echo  dance   

               city      time        day  
0  Saint-Petersburg  20:28:33  Wednesday  
1            Moscow  14:07:09     Friday  
2  Saint-Petersburg  20:58:07  Wednesday  
3  Saint-Petersburg  08:37:09     Monday  
4            Moscow  08:34:34     Monday

In [14]:
df.groupby('city')['genre'].count()

city
Moscow              42741
Saint-Petersburg    18512
Name: genre, dtype: int64

In [15]:
df.groupby('day')['genre'].count()

day
Friday       21840
Monday       21354
Wednesday    18059
Name: genre, dtype: int64

In [16]:
def num_of_listenings(day, city):
    track_list = df[(df['city'] == city) & (df['day'] == day)]
    track_list_count = track_list['genre'].count()
    return track_list_count

In [17]:
num_of_listenings('Monday', 'Moscow')

15740

In [18]:
num_of_listenings('Monday', 'Saint-Petersburg')

5614

In [19]:
num_of_listenings('Wednesday', 'Moscow')

11056

In [20]:
num_of_listenings('Wednesday', 'Saint-Petersburg')

7003

In [21]:
num_of_listenings('Friday', 'Moscow')

15945

In [22]:
num_of_listenings('Friday', 'Saint-Petersburg')

5895

In [23]:
columns = ['city','monday','wednesday','friday']
data = [['Moscow','15740','11056','15945'],
        ['Saint-Petersburg','5614','7003','5895']]
listenings_statistic = pd.DataFrame(data = data, columns = columns)
listenings_statistic

city monday wednesday friday
0            Moscow  15740     11056  15945
1  Saint-Petersburg   5614      7003   5895

Пик прослушиваний в Москве по понедельникам и пятницам, а в Санкт-Петербурге напротив - спад пн. и пт. и пик в среду. Гипотеза верна.

## Утром в понедельник в Москве преобладают одни жанры музыки, а в Петербурге — другие. Это верно и для вечера пятницы.

In [24]:
df_moscow = df[df['city'] == 'Moscow']
df_spb = df[df['city'] == 'Saint-Petersburg']
print(df_moscow.head())
df_spb.head()

    user_id                        track            artist   genre    city  \
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock  Moscow   
4  E2DC1FAE                  Soul People        Space Echo   dance  Moscow   
6  4CB90AA5                         True      Roman Messer   dance  Moscow   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance  Moscow   
8  8FA1D3BE     И вновь продолжается бой           unknown  ruspop  Moscow   

       time        day  
1  14:07:09     Friday  
4  08:34:34     Monday  
6  13:00:07  Wednesday  
7  20:47:49  Wednesday  
8  09:17:40     Friday  


user_id                  track            artist   genre  \
0  FFB692EC      Kamigata To Boots  The Mass Missile    rock   
2    20EC38      Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9  Dragons in the Sunset        Fire + Ice    folk   
5  842029A1              Преданная         IMPERVTOR  rusrap   
9  E772D5C0              Pessimist           unknown   dance   

               city      time        day  
0  Saint-Petersburg  20:28:33  Wednesday  
2  Saint-Petersburg  20:58:07  Wednesday  
3  Saint-Petersburg  08:37:09     Monday  
5  Saint-Petersburg  13:09:41     Friday  
9  Saint-Petersburg  21:20:49  Wednesday

In [25]:
def weekdays_genres(time1, time2, day, df_city):
    df_genre = df_city[df_city['time'] >= time1]
    df_genre = df_city[df_city['time'] <= time2]
    df_genre = df_genre[df_genre['day'] == day]
    df_genre = df_genre.groupby('genre')['genre'].count()
    df_genre_sort = df_genre.sort_values(ascending = False)
    return df_genre_sort.head(10)

In [26]:
weekdays_genres('6:00', '12:00', 'Monday', df_moscow)

genre
pop            781
dance          549
electronic     480
rock           474
hiphop         286
ruspop         186
world          181
rusrap         175
alternative    164
unknown        161
Name: genre, dtype: int64

In [27]:
weekdays_genres('6:00', '12:00', 'Monday', df_spb)

genre
pop            218
dance          182
rock           162
electronic     147
hiphop          80
ruspop          64
alternative     58
rusrap          55
jazz            44
classical       40
Name: genre, dtype: int64

In [28]:
weekdays_genres('6:00', '12:00', 'Friday', df_moscow)

genre
pop            789
dance          585
rock           509
electronic     506
hiphop         277
ruspop         207
classical      201
world          174
alternative    172
rusrap         148
Name: genre, dtype: int64

In [29]:
weekdays_genres('6:00', '12:00', 'Friday', df_spb)

genre
pop            211
dance          192
electronic     167
rock           156
hiphop         109
classical       56
rusrap          55
alternative     55
world           46
ruspop          45
Name: genre, dtype: int64

Из полученных результатов видно, что топ 10 жанров в обоих городах занимают одни и те же жанры. Гипотеза не подтвердилась, однако из-за пропусков в изначальных данных в понедельник в Москве жанр 'unknown'(пропуск значения в .csv) занял 10 место. Пропуски ставят результат под сомнения.

## Москва и Петербург предпочитают разные жанры музыки. В Москве чаще слушают поп-музыку, в Петербурге — русский рэп.

In [30]:
df_msk_genres = df[df['city'] == 'Moscow']
df_msk_genres = df_msk_genres.groupby('genre')['genre']
df_msk_genres = df_msk_genres.count()
df_msk_genres = df_msk_genres.sort_values(ascending = False)
df_msk_genres.head(10)

genre
pop            5892
dance          4435
rock           3965
electronic     3786
hiphop         2096
classical      1616
world          1432
alternative    1379
ruspop         1372
rusrap         1161
Name: genre, dtype: int64

In [31]:
df_spb_genres = df[df['city'] == 'Saint-Petersburg']
df_spb_genres = df_spb_genres.groupby('genre')['genre']
df_spb_genres = df_spb_genres.count()
df_spb_genres = df_spb_genres.sort_values(ascending = False)
df_spb_genres.head(10)

genre
pop            2431
dance          1932
rock           1879
electronic     1736
hiphop          960
alternative     649
classical       646
rusrap          564
ruspop          538
world           515
Name: genre, dtype: int64

In [32]:
df.groupby('city')['genre'].count()

city
Moscow              42741
Saint-Petersburg    18512
Name: genre, dtype: int64

In [33]:
columns = ['city', 'percentage_rock', 'percentage_pop']
data = [['Moscow',42741/3965, 42741/5892],
        ['Saint-Petersburg', 18512/1879, 18512/2431]]
pd.DataFrame(data = data, columns = columns)

city  percentage_rock  percentage_pop
0            Moscow        10.779571        7.254073
1  Saint-Petersburg         9.852049        7.614973

In [34]:
df_genre = df.groupby('genre')['genre'].count().sort_values(ascending = False)
df_genre.head(15)

genre
pop            8323
dance          6367
rock           5844
electronic     5522
hiphop         3056
classical      2262
alternative    2028
world          1947
ruspop         1910
rusrap         1725
jazz           1466
metal          1210
unknown        1127
soundtrack     1116
folk            972
Name: genre, dtype: int64

Как мы видим из таблицы, в процентном соотношении в Москве чаще слушают рок, чем в СПБ, а в СПБ чаще слушают поп, чем в Москве. Гипотеза не подтвердилась, однако стоит учитывать, что значения 'unknown' могли бы изменить статистику => таких значений 1127.